[#74] 내 동네 삭제 기능 추가

Srltas opened the pull request:
## 📌 Summary
사용자가 내 동네를 삭제하는 기능을 추가합니다.

## 📝 Description
- 동네 ID를 받아 내 동네를 삭제하는 기능 구현

**내 동네 삭제 API**
- 메서드: DELETE
- 엔드포인트: /users/neighborhoods/{neighborhoodId}

## ✅ Checklist
- [x] 새로운 기능이나 수정된 기능에 대해 충분한 테스트를 작성했습니다.
- [x] 코딩 스타일 가이드를 준수했습니다.
- [x] 문서(주석, README 등)가 필요하다면 업데이트했습니다.

Srltas committed:
feat: User 도메인 내 동네 삭제 기능 추가

Srltas committed:
feat: 내 동네 삭제 Service 구현

Srltas committed:
feat: 내 동네 삭제 Repository 구현

Srltas committed:
feat: 내 동네 삭제 Controller 구현

Srltas committed:
test: 내 동네 삭제 기능 테스트 코드 추가

f-lab-bradley commented on src/main/java/com/srltas/runtogether/adapter/in/UserNeighborhoodController.java:
```diff
@@ -21,23 +26,41 @@
 import lombok.RequiredArgsConstructor;
 
 @Tag(name = "내 동네 API", description = "내 동네 API")
+@RequestMapping(USER_NEIGHBORHOOD_REGISTRATION)
```
_REGISTRATION
변수 명을 이렇게 설정 했는데 아래 delete 에서는 삭제를 진행 하고 있네요.
변수 명을 변경 하는건 어떨까요?

Srltas replied:
잘못된 변수명을 사용한거 같습니다!..
USER_NEIGHBORHOOD로 변경하는 것이 적절한거 같아 수정했습니다.

f-lab-bradley commented on src/main/java/com/srltas/runtogether/adapter/in/UserNeighborhoodController.java:
```diff
@@ -21,23 +26,41 @@
+		summary = "내 동네 삭제",
+		description = "사용자의 내 동네를 삭제합니다."
+	)
+	@ApiResponse(responseCode = "200", description = "내 동네 삭세 성공")
```
삭세 -> 삭제

f-lab-bradley commented on src/main/java/com/srltas/runtogether/adapter/in/web/common/UrlConstants.java:
```diff
@@ -8,10 +8,11 @@
 @UtilityClass
 public class UrlConstants {
 	public final String ALL_URL = "/*";
+	public final String USER_NEIGHBORHOOD_PATTERN = "/users/neighborhoods/*";
```
여기 사용되는 URL 은 Controller 와 이어지는 것이 좋겠어요.

Srltas replied:
Controller에서 사용하는 URL과 Filter에서 사용하는 URL을 나눠서 따로 관리하는 것이 좋을거 같습니다.
- ControllerUrls: Controller에 매핑될 URL
- FilterUrls: 사용자 인증, 인가가 필요한 URL

이 변경사항은 새로운 이슈[#82]로 생성해 수정하도록 하겠습니다.

f-lab-bradley commented on src/main/java/com/srltas/runtogether/adapter/out/persistence/mybatis/MybatisUserRepositoryLogProxy.java:
```diff
@@ -39,6 +39,13 @@ public void addUserNeighborhood(String userId, String neighborhoodId) {
+	public void deleteUserNeighborhood(String userId, String neighborhoodId) {
+		long startTime = System.currentTimeMillis();
+		userRepository.deleteUserNeighborhood(userId, neighborhoodId);
+		RunTogetherMDC.putMessage(DATABASE_FETCH_TIME, String.valueOf(System.currentTimeMillis() - startTime));
```
조회 후 삭제 하는 경우를 생각 해 보면 Database fetch time 이 중복 될 수 있을 것 같아요.
특정 쿼리를 특정 할 수 있는 네이밍 방식이 좋을 것 같아요.

Srltas replied:
로그 이름이 중복으로 남는 문제가 발생할 수 있겠네요!..
수행되는 쿼리를 특정할 수 있는 이름을 부여하도록 하겠습니다.

f-lab-bradley commented on src/main/java/com/srltas/runtogether/application/UserNeighborhoodService.java:
```diff
@@ -30,4 +32,11 @@ public void addNeighborhood(AddUserNeighborhoodCommand command) {
+
+	@Override
+	public void deleteUserNeighborhood(DeleteUserNeighborhoodCommand command) {
+		User user = userRepository.findById(command.userId()).orElseThrow(UserNotFoundException::new);
```
이건 그냥 선택인데요, 
저라면 `user 를 조회하고 동네를 삭제한다` 를 도메인 서비스에 넣을 것 같아요.
유즈케이스에서는 그저 
`사용자의 동네를 삭제 한다.` 정도로만 남기구요.
어느정도 까지가 도메인레이어에서 처리해야 하는 핵심 비지니스 로직인지를 고민 해 보면 좋을 것 같아요.
근데 이건 크게 중요한건 아님

Srltas replied:
`user를 조회하고 동네를 삭제` 로직을 처리하는 새로운 클래스를 생성하는 것인가요?

f-lab-bradley commented on src/main/java/com/srltas/runtogether/application/UserNeighborhoodService.java:
```diff
@@ -30,4 +32,11 @@ public void addNeighborhood(AddUserNeighborhoodCommand command) {
+	@Override
+	public void deleteUserNeighborhood(DeleteUserNeighborhoodCommand command) {
+		User user = userRepository.findById(command.userId()).orElseThrow(UserNotFoundException::new);
+		user.deleteNeighborhood(command.neighborhoodId());
```
deleteNeighborhood 에서 NeighborhoodNotRegisteredException 을 던지는데 이 유즈케이스에선 굳이 필요한 에러 일까요??
도메인에서 해당 에러를 던진다고 하더라도 use case 에서는 `원래 없던걸 지우라는 요청을 했으니 어쨌든 결과적으로는 지워짐.` 으로 봐도 되지 않을까요???
해당 에러를 잡아서 아무것도 하지 않고 성공으로 내리는 방식도 고려를 해 보시면 좋을 것 같아요.

Srltas replied:
원래 없던 것을 지우라는 요청을 받았으니, 결과적으로 삭제된 셈이라는 새로운 관점도 흥미롭습니다. 결론적으로 내가 등록한 동네가 존재하지 않는 상태가 맞으므로, 사용자에게 성공으로 처리해도 괜찮을 같습니다!

다만, 해당 상황에 대한 로그는 남기는 것이 좋다고 생각해 예외 발생 시 로그를 기록하되, 사용자에게는 성공으로 표시하도록 수정했습니다.

f-lab-bradley reviewed (commented):
하나의 API 를 한 번에 PR 올리기 보다는 각 Layer 에 맞게 쪼개서 PR 을 올리는 것도 생각 해 보시면 좋을 것 같아요~

Srltas committed:
dosc: 동네 삭제 오타 수정

Srltas committed:
refactor: 내 동네 컨트롤러 URL 상수 공통된 이름으로 변경

Srltas committed:
refactor: URL 와일드카드 매칭 처리 기능 추가

Srltas committed:
refactor: SQL 별로 log 필드 이름 설정

Srltas committed:
refactor: 내 동네 삭제 예외 처리 규칙 변경

- 내 동네로 등록되어 있지 않은 동네를 삭제하는 경우, 결과적으로 내 동네에 없는 상태이다.
- 따라서, 사용자에게는 성공으로 반환하고 예외에 대해서는 로그로 기록해 남긴다.

Srltas reviewed (commented):
처음에는 코드의 변경이 많이 없을 것이라 생각해 한 번에 작업을 진행 했지만, 잘못된 생각이었습니다..
다음 작업에서는 다시 세분화 해서 PR을 만들도록 하겠습니다!..
